src/server/pfs/server/reader.go

```diff
@@ -0,0 +1,263 @@
+	if err := fr.fmr.Get(w); err != nil {
+		return err
+	}
+	for fr.fileCount > 0 {
+		fmr, err := fr.mr.Next()
+		if err != nil {
+			return err
+		}
+		if err := fmr.Get(w); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	if len(noPadding) > 0 && noPadding[0] {
+		return nil
+	}
+	// Close a tar writer to create tar EOF padding.
+	return tar.NewWriter(w).Close()
+}
+
+func (fr *FileReader) drain() error {
+	for fr.fileCount > 0 {
+		if _, err := fr.mr.Next(); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	return nil
+}
+
+// Reader iterates over FileInfoV2s generated from a fileset.ReaderAPI
+type Reader struct {
+	commit    *pfs.Commit
+	getReader func() (fileset.FileSource, error)
+}
+
+// Creates a Reader which emits FileInfoV2s with the information from commit, and the entries from readers
+// returned by getReader.  If getReader returns different Readers all bets are off.
+func NewReader(commit *pfs.Commit, getReader func() (fileset.FileSource, error)) *Reader {
+	return &Reader{
+		commit:    commit,
+		getReader: getReader,
+	}
+}
+
+func (r *Reader) Iterate(ctx context.Context, cb func(*pfs.FileInfoV2, fileset.File) error) error {
+	r1, err := r.getReader()
+	if err != nil {
+		return err
+	}
+	r2, err := r.getReader()
+	if err != nil {
+		return err
+	}
+	s1, s2 := newStream(r1), newStream(r2)
+	defer s1.Close()
+	defer s2.Close()
+	cache := make(map[string][]byte)
+	for {
+		fr, err := s1.Next()
+		if err != nil {
+			return err
+		}
+		if fr == nil {
+			return nil
+		}
+		idx := fr.Index()
+		hashBytes, err := computeHash(cache, s2, idx.Path)
+		if err != nil {
+			return err
+		}
+		finfo := &pfs.FileInfoV2{
+			File: client.NewFile(r.commit.Repo.Name, r.commit.ID, idx.Path),
+			Hash: pfs.EncodeHash(hashBytes),
+		}
+		if err := cb(finfo, fr); err != nil {
+			return err
+		}
+		delete(cache, idx.Path)
+	}
+}
+
+func computeHash(cache map[string][]byte, s *stream, x string) ([]byte, error) {
+	if hashBytes, exists := cache[x]; exists {
+		return hashBytes, nil
+	}
+	// check that the next item in the stream is x
+	idx := s.Peek()
+	if idx == nil {
+		return nil, errors.Errorf("stream is done, can't compute hash for %s", x)
+	}
+	if idx.Path != x {
+		return nil, errors.Errorf("stream is is wrong place to compute hash for %s", x)
+	}
+	// consume x from the stream
+	fr, err := s.Next()
+	if err != nil {
+		return nil, err
+	}
+	idx = fr.Index()
+	h := pfs.NewHash()
+	if indexIsDir(idx) {
+		// for directory
+		for {
+			idx2 := s.Peek()
+			if idx2 == nil {
+				break
+			}
+			if !strings.HasPrefix(idx2.Path, x) {
+				break
+			}
+			childHash, err := computeHash(cache, s, idx2.Path)
+			if err != nil {
+				return nil, err
+			}
+			h.Write(childHash)
+		}
+	} else {
+		// for file
+		if idx.DataOp != nil {
+			for _, dataRef := range idx.DataOp.DataRefs {
+				h.Write([]byte(dataRef.Hash))
+			}
+		}
+	}
+	hashBytes := h.Sum(nil)
+	cache[x] = hashBytes
+	return hashBytes, nil
+}
+
+type stream struct {
```

Comment:
I think the following would be a cleaner design for the stream (and is closer to the stream I implemented for the datum stream). We could add a context that would be used for breaking out of the iteration so we don't need an explicit close that forces the full iteration. Also, `io.EOF` is a better way to indicate the end of  the stream rather than a nil file returned.

```
type stream struct {
	peek     fileset.File
	fileChan chan fileset.File
	errChan  chan error
}

func newStream(source fileset.Source) *stream {
	fileChan := make(chan fileset.File)
	errChan := make(chan error, 1)
	go func() {
		if err := source.Iterate(func(file fileset.File) error {
                        fileChan <- file
			return nil
		}); err != nil {
			errChan <- err
			return
		}
		close(fileChan)
	}()
	return &stream{
		fileChan: fileChan,
		errChan:  errChan,
	}
}

func (s *stream) Peek() (fileset.File, error) {
        if s.peek != nil {
            return s.peek, nil
        }
        var err error
        s.peek, err = s.Next()
        return s.peek, err
}

func (s *stream) Next() (fileset.File, error) {
        if s.peek != nil {
            tmp := s.peek
            s.peek = nil
            return tmp, nil
        }
	select {
	case file, more := <-s.fileChan:
		if !more {
			return nil, io.EOF
		}
		return file, nil
	case err := <-s.errChan:
		return nil, err
	}
}
```